lib/codecs/src/decoding/format/vrl.rs

```diff
@@ -0,0 +1,293 @@
+                .to_string()
+                .into()),
+        }
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                schema::Definition::empty_legacy_namespace().unknown_fields(Kind::any())
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::any(), [log_namespace])
+            }
+        }
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing logs compatible with VRL.
+#[derive(Debug, Clone)]
+pub struct VrlDeserializer {
+    program: Program,
+    timezone: TimeZone,
+}
+
+fn parse_bytes(bytes: Bytes, log_namespace: LogNamespace) -> Event {
+    let bytes_deserializer = BytesDeserializerConfig::new().build();
+    let log_event = bytes_deserializer.parse_single(bytes, log_namespace);
+    Event::from(log_event)
+}
+
+impl Deserializer for VrlDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let event = parse_bytes(bytes, log_namespace);
+        match self.run_vrl(event, log_namespace) {
+            Ok(event) => Ok(smallvec![event]),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl VrlDeserializer {
+    fn run_vrl(&self, event: Event, log_namespace: LogNamespace) -> vector_common::Result<Event> {
+        let mut runtime = Runtime::default();
+        let mut target = VrlTarget::new(event, self.program.info(), true);
+        match runtime.resolve(&mut target, &self.program, &self.timezone) {
+            Ok(_) => match target.into_events(log_namespace) {
+                TargetEvents::One(event) => Ok(event),
+                TargetEvents::Logs(_) => Err("multiple targets are not supported".into()),
+                TargetEvents::Traces(_) => Err("trace targets are not supported".into()),
+            },
+            Err(e) => Err(e.to_string().into()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use chrono::{DateTime, Utc};
+    use indoc::indoc;
+    use vrl::btreemap;
+    use vrl::path::OwnedTargetPath;
+    use vrl::value::Value;
+
+    fn make_decoder(source: &str) -> VrlDeserializer {
+        VrlDeserializerConfig {
+            vrl: VrlDeserializerOptions {
+                source: source.to_string(),
+                timezone: None,
+            },
+        }
+        .build()
+        .expect("Failed to build VrlDeserializer")
+    }
+
+    #[test]
+    fn test_json_message() {
+        let source = indoc!(
+            r#"
+            . = string!(.)
+            . = parse_json!(.)
```

Comment:
Is setting metadata also supported? E.g. `%foo = "bar"`.

Reply:
Yes, this is supported. Added metadata to a test.